Islandora 1889 Escapeing search terms, preventing error searching solr index.

MorganDawe opened the pull request:
**JIRA Ticket**: (https://jira.duraspace.org/browse/ISLANDORA-1889)

# What does this Pull Request do?
When preforming an advanced search on the islandora advanced search block, special characters (such as '[' or ']') will throw an 'Error searching Solr index "400" Status: Bad Request'.

# What's new?
Implements the [islandora_solr_facet_escape()|https://github.com/Islandora/islandora_solr_search/blob/7.x/includes/utilities.inc#L165] utility function on the search term prior to query execution.

# How should this be tested?
- Navigate to the advanced search block.
- Preform a search query with an invalid character (such as 'Lobster[' or 'lobster{{}'). Results pages should throw a 400 error.

# Steps to reproduce
- Navigate to the advanced search block.
- Preform a search query with an invalid character (such as 'Lobster[' or 'lobster{{}'). Results pages should not throw a 400 error.

# Interested parties
@Islandora/7-x-1-x-committers

MorganDawe committed:
escapeing search terms.

willtp87 commented:
I'm seeing a search with the special characters result with the form value on the next page having the back slashes.

MorganDawe committed:
Correcting escaped string passed to form when preserving filters.

MorganDawe committed:
Removed debug code.

MorganDawe committed:
Changed escape function to utilize lesser escape instead.

MorganDawe committed:
Updated current query with stripslashes on query.

MorganDawe committed:
Refactored to allow for configuration of escapable fields.

MorganDawe committed:
General purpose clean up.

rosiel commented:
Don't forget to uninstall your new variables! (this pull and https://github.com/Islandora/islandora_solr_facet_pages/pull/32)

MorganDawe committed:
Added variables to uninstall hook.

MorganDawe commented:
Thanks @rosiel 👍

MorganDawe committed:
Added module name to variables.

MorganDawe committed:
updated variable get names in form.

MorganDawe committed:
Correcting some syntax issues.

adam-vessey commented on islandora_solr.install:
```diff
@@ -53,6 +53,8 @@ function islandora_solr_uninstall() {
   $variables = array_merge(islandora_solr_search_settings_variables(), array(
     'islandora_solr_url',
+    'islandora_solr_advanced_search_block_lucene_syntax_escape',
+    'islandora_solr_advanced_search_block_lucene_charset_default,',
```
Looks like there's a comma as the last character of the variable name, here... It was probably supposed to be outside the single-quotes originally,

DiegoPino reviewed (changes requested):
As I stated here also https://github.com/Islandora/islandora_solr_facet_pages/pull/32#issuecomment-271287300

I don't think its a good idea to add Solr reserved characters escaping strategies and configs all over the place. We should have just one, just a single escaping function. Also giving users the option to define which characters should be considered is confusing because Solr ones are pretty well documented, they don't change and any user contributed small change on that config field will result in a lot, hard to debug errors for a final user/web admin not involved in the code itself.
I would prefer a simpler, unified, shared with facets, etc, strategy. What is the use case for having user configurable special characters?

MorganDawe committed:
Update islandora_solr.install

whikloj commented on includes/blocks.inc:
```diff
@@ -694,6 +694,10 @@ function islandora_solr_advanced_search_form_submit($form, &$form_state) {
     if (!empty($search)) {
+      if (variable_get('islandora_solr_advanced_search_block_lucene_syntax_escape', TRUE)) {
+        $escape_charset = variable_get('islandora_solr_advanced_search_block_lucene_charset_default', "+,-,&&,||,!,(,),{,},[,],^,~,:");
+        $search = islandora_solr_facet_query_escape($search, explode(',', $escape_charset));
```
Why not just use [islandora_solr_lesser_escape](https://github.com/Islandora/islandora_solr_search/blob/7.x/includes/utilities.inc#L139) or [islandora_solr_facet_escape](https://github.com/Islandora/islandora_solr_search/blob/7.x/includes/utilities.inc#L165).

Either of these would resolve the issue in the JIRA ticket and then there is no chance a user could remove a needed character. Or is there a use case for an expanded filter that I missed?

MorganDawe commented:
@whikloj , @DiegoPino @Islandora/7-x-1-x-committers,
 
The idea here is to avoid Solr errors when performing a Solr search. For example, a user may wish to search for an object with a title of ["a nurse by her station"] (Including the '"', '[' and ']'). Currently, this search will throw a 400 error. Of course, manually escaping the search string properly will work, however the average user performing a search may not know that. It was suggested perhaps using the 'islandora_solr_facet_escape()' utility function, however that can't really be justified because it is entirely reasonable to search for a quoted phrase. It's about appealing to a user base and it's comfort level and familiarity with string escaping and apache Solr Lucene syntax.

To be sure, we must consider two use cases for user search and discovery. 

- When a user wishes to search for an exact phrase in quotations, such as "search text"

- When a user wishes to search for content with a special character, but is unfamiliar with escaping functionality, such as ["a nurse by her station"].

It is also very common to see queries with a '+' or '-', which would currently be escaped if using 'islandora_solr_lesser_escape()'.

The escape character list has been exposed as configurable, as a set of reasonable defaults based on the current Escaping Special Characters list of: + - && || ! ( ) { } [ ] ^ " ~ * ? : \. It may also be reasonable for an institution to limit a given escaping character set for the sake of simplicity, or to avoid potential collisions given an institution's metadata content. 

Because multiple use cases are presenting themselves, it would appear reasonable to expose this as configuration. Forcing organizations into using 'islandora_solr_facet_escape()' or 'islandora_solr_lesser_escape()', forces all into one methodology or the other. Also, facet escaping advanced search terms could result in unexpected results (see first use case).

Furthermore, I would like to add that I agree with the thought of simplicity in configuration. Perhaps Solr facet pages and advanced search SHOULD have Solr Lucene escape character syntax configured together in one spot, but configuration for one may not be desirable for another.

I thank everybody involved for their contribution.

whikloj commented:
@MorganDawe If I went in and changed my escape characters from the default to `+, -` and then someone does the suggested searches of `Lobster[` or `Lobster{{}`. Solr would throw a 400, right? So why allow that to happen?

What you are describing seems to be a custom search interface for users who need access to specific syntax. Say a metadata librarian that wants to search `"betty ford"{person}` or you want to pass actual Solr syntax `{!join from=ancestors_ms to=PID}`.

Both of which is are valid use cases and I see the benefit of customizability you are suggesting, I'm just saying that this is an easy way for people to break their site. That's why it seems like this would be better as a custom search block.

willtp87 commented:
@MorganDawe  it looks like you have regressed on re-escaping searches. Also, I think double quotes should be in the default list.

MorganDawe committed:
Fixing some revision, updated form, and default regex value.

MorganDawe committed:
Merge branch '7.x-1889' of https://github.com/MorganDawe/islandora_solr_search into 7.x-1889

MorganDawe committed:
Removed some test code.

MorganDawe committed:
Updating logic on toggle.

willtp87 commented on includes/blocks.inc:
```diff
@@ -694,6 +698,14 @@ function islandora_solr_advanced_search_form_submit($form, &$form_state) {
+        $search = islandora_solr_lesser_escape($search);
+        $escape_charset = variable_get(
+          'islandora_solr_advanced_search_block_lucene_charset_default',
+          '+,-,&&,||,!,(,),{,},[,],^,",~,*,?,:,\"'
```
Not consistent, we should move to a constant.

willtp87 replied:
I think all the special chars should be in this list including `\`.

willtp87 commented on includes/results.inc:
```diff
@@ -202,7 +202,13 @@ class IslandoraSolrResults {
     if (!in_array($islandora_solr_query->solrQuery, $islandora_solr_query->differentKindsOfNothing)) {
       // Get query value.
-      $query_value = $islandora_solr_query->solrQuery;
+      if (variable_get('islandora_solr_advanced_search_block_lucene_syntax_escape', TRUE)) {
```
The var should probably default to FALSE.

MorganDawe committed:
adding regex for testing.

willtp87 commented on includes/utilities.inc:
```diff
@@ -118,6 +118,57 @@ function islandora_solr_csv_explode($delim = ',', $str = '', $enclose = '"', $pr
+      '"',
+    );
+  }
+  foreach ($chars as $char) {
```
This iteration can lead to trouble; it should probably be using preg_repalce.

MorganDawe committed:
Updated with regex info.

MorganDawe committed:
Merge pull request #2 from MorganDawe/7.x-regex

7.x regex

DiegoPino commented on includes/admin.inc:
```diff
@@ -529,6 +529,25 @@ function islandora_solr_admin_settings($form, &$form_state) {
+  $form['islandora_solr_tabs']['advanced_search_block']['islandora_solr_advanced_search_block_lucene_regex_default'] = array(
+    '#type' => 'textfield',
+    '#title' => t('Default regular expression evaluated on search term'),
+    '#default_value' => variable_get('islandora_solr_advanced_search_block_lucene_regex_default', '/(\+|-|&&|\|\||!|\(|\)|\{|\}|\[|\]|\^|~|\*|\?|\:|\")/'),
```
Would love to see this  being a CONSTANT so other solution packs (like facet pages) could be more consistent if possible

MorganDawe replied:
Had that thought as well, will do.

MorganDawe committed:
Added const lucene facet escape character regex.

MorganDawe committed:
Merge pull request #3 from MorganDawe/7.x-regex

Added const lucene facet escape character regex.

DiegoPino commented:
@MorganDawe almost there. Testing with real content. Will merge soon. thanks!

whikloj commented:
@MorganDawe I'm sorry I just realized I misread this.

I thought that if you did **not** enable this toggle then your searches would get escaped by the  `islandora_solr_lesser_escape` or `islandora_solr_facet_escape` (or even maybe the new `const` you created). Then if you **did** enable it, you could configure what was escaped.

I know that people will find the default search behaviour changes, but it is currently broken as defined in the ticket so I would think that by default people should get escaping and people can choose to remove/alter the escaped characters if they want.

Also @DiegoPino brought up a good question, if someone changes their configuration they may not remember how to change it back. Perhaps we need a reset button or at least list the default string (again that `const`) so people could copy/paste?

Sorry, I know you've put a lot of time into this. What are your thoughts?

DiegoPino commented:
@MorganDawe I tested and in indeed it is being escaped correctly. But I do have some  questions still
A: defaults

1. Should escaping via internal, preset and defined by the CONSTANT character list, regular expression pattern,  the default strategy, assuming right now it's broken and that is the expected way
2. The optional alternative being enabling a custom pattern, instead of the default?

B: About consistency. I saw that by enabling escaping breadcrumbs changed for special characters, e.g 
an object with an "//" in the title. Which displays fine, it is there, I can search for that (with some tunning) directly on Solr via `q=fgs_label_s:*//*`

If enabling escaping I get `~slsh~` replacements in the field, in the query and in the breadcrumb which traduces to 0 results and raw query with 
````
"debug":{
    "rawquerystring":"fgs_label_s:(\\*\\~slsh\\~\\~slsh\\~\\*)",
    "querystring":"fgs_label_s:(\\*\\~slsh\\~\\~slsh\\~\\*)",
    "parsedquery":"fgs_label_s:*~slsh~~slsh~*",
    "parsedquery_toString":"fgs_label_s:*~slsh~~slsh~*",
````
if not, I get `~slsh~`on the query, but  breadcrumb stays "//" and also search field and in all cases I get 1 search results which is fine

```` debug":{
    "rawquerystring":"fgs_label_s:(*//*)",
    "querystring":"fgs_label_s:(*//*)",
    "parsedquery":"fgs_label_s:*//*",
    "parsedquery_toString":"fgs_label_s:*//*",
````

Is that OK? Wondering if escaping strategy is correct if i really need to search fo content with a // (imagine I'm searching for code comments!)

Thanks

MorganDawe commented:
@DiegoPino @whikloj Good catch, I think i made a silly mistake [here](https://github.com/Islandora/islandora_solr_search/pull/301/files#diff-3c926c7f5c8cd595044032e599489b9fR707). Should have been changed with the recent refactor that utilized the regular expression pattern. As it is now, the regular expression pattern is being exploded as an array. If your going to explode regular expressions on comma's, your going to have a bad time. I need to fix this, and i will get to it very shortly.

The presence of ~slsh~ should be coming from 'islandora_solr_replace_slashes()' and removed with 'islandora_solr_remove_slashes()'. I suspect the previously mentioned bug correction will help resolve this issue, so i will test further to ensure it gets corrected.

Its only my opinion, but i feel it would likely be best to keep this functionality off by default. Atleast for now. Since turning it on by default would propagate the need for heavy documentation around the usage of advanced search (or facet pages search). 

Currently, field descriptions appear with the textbox (where the regular expression can be modified) that indicate what the default value is (which is using the ISLANDORA_SOLR_QUERY_FACET_LUCENE_ESCAPE_REGEX const). Would you suggest perhaps a tool tip to accompany the description as well? I think a reset button would  also work well here to btw.

Thank you both for taking the time to test this out 👍

MorganDawe committed:
Cleaning up some potential issues with strip slashes and array explosion.

MorganDawe committed:
Cleaning code for review.

MorganDawe committed:
Merge pull request #4 from MorganDawe/7.x-regex

7.x regex

MorganDawe commented:
@DiegoPino @whikloj i did some further troubleshooting, and it turns out a large issue was due to the replace slashes string added and removed by islandora_solr_replace_slashes(). Because islandora_solr_facet_query_escape() was being called after islandora_solr_replace_slashes(), the special characters in the replacement pattern were also being escaped (such as `~slsh~` and `~plus~`), due to the `~` character appearing in the ISLANDORA_SOLR_QUERY_FACET_LUCENE_ESCAPE_REGEX constant by default. Some refactoring was done in the advanced search form submit to allow this.

As you know, some characters cannot appear in a query string when present in the url such as '/' and '+', as it will break the url essentially. This is why it was necessary to preserve this functionality.

I have also confirmed that the replacement pattern appears in the form textfield description (so one may copy/paste it back as a default) and added the newly created variables     'islandora_solr_advanced_search_block_lucene_syntax_escape' and 'islandora_solr_advanced_search_block_lucene_regex_default' to the list of variables to be reset when clicking 'Reset to defaults' on the SOLR settings admin page.

Your knowledgeable feedback is invaluable to me, so thank you very much in advance for any further issues you may uncover or suggestions for further improvements :).

adam-vessey commented on includes/blocks.inc:
```diff
@@ -684,16 +690,41 @@ function _islandora_solr_advanced_search_terms($form, &$form_state) {
+        // Replace the occurances of '~slsh~' and '~plus~' after
+        // islandora_solr_facet_query_escape(), since they both contain
+        // '~', which would be improperlly escaped prior to query execution.
+        $search = islandora_solr_replace_slashes($search);
```
Instead of hitting `islandora_solr_replace_slashes()` two places as it does now, why not move it down to immediately before adding to the `$query_array` below?... Could then have just the one call...

... Really should almost go lower, when defining the `$form_state['redirect']`, on all of `$query`, so we hit any funky "base" queries which might be defined.

adam-vessey replied:
~~... Looking a little further, maybe not down to the query bit, due to how it does the field separate from the terms... Will need two calls to replace: one for the field, the second for the terms (not two for the terms based on another condition).~~ I was running into browser issues... (Backslashes are handled oddly; passed back URL-escaped (expected), Firefox displays as an actual backslash (unexpected), and does some weird things because of it).

adam-vessey commented on includes/blocks.inc:
```diff
@@ -556,13 +555,20 @@ function islandora_solr_advanced_search_form($form, &$form_state) {
   foreach ($values['terms'] as $i => $value) {
+    if (variable_get('islandora_solr_advanced_search_block_lucene_syntax_escape', FALSE)) {
+      if (isset($value['search'])) {
+        $value['search'] = islandora_solr_restore_slashes($value['search']);
```
I don't believe this should actually be necessary, assuming the query has actually been built... [One of the first things the query processor does is to run restore slashes](https://github.com/Islandora/islandora_solr_search/blob/b185566d4aa1d9cc28b68823988affea5c742f61/includes/query_processor.inc#L121).

adam-vessey commented on includes/blocks.inc:
```diff
@@ -556,13 +555,20 @@ function islandora_solr_advanced_search_form($form, &$form_state) {
     '#tree' => TRUE,
   );
   foreach ($values['terms'] as $i => $value) {
+    if (variable_get('islandora_solr_advanced_search_block_lucene_syntax_escape', FALSE)) {
```
Given we're replacing slashes independent of this... should probably be restoring independently?

adam-vessey committed:
Replace once.

adam-vessey committed:
Somewhat more consistent?

adam-vessey committed:
Rename the constant and variable(s).

adam-vessey committed:
On second thought, let's leave the original constants as originally defined.

... And not move to `const`.

MorganDawe committed:
Merge pull request #5 from adam-vessey/7.x-ISLANDORA-1889

Adjustments...

MorganDawe commented:
Working with @adam-vessey, as he had some contributions to make. A pull issued to my repo has been merged and is thus included here. Thanks @adam-vessey for his hard work and contributions 👍

DiegoPino commented:
@MorganDawe ok, cool. Give me a ping once you feel it's ready for testing again. thanks!

MorganDawe committed:
øOMerge branch '7.x-1889' of https://github.com/MorganDawe/islandora_solr_search into MorganDawe-7.x-1889

MorganDawe committed:
Updating const name appropriately.

MorganDawe committed:
Merge pull request #6 from MorganDawe/MorganDawe-7.x-1889

Morgan dawe 7.x 1889

MorganDawe commented:
@DiegoPino This should be ready for another review, i apologize for the late reply.

MorganDawe commented:
@Islandora/7-x-1-x-committers for review

DiegoPino commented:
@MorganDawe will review today. Thanks

MorganDawe commented:
@DiegoPino Thank you for offering to review this pull.

DiegoPino commented:
@MorganDawe almost ready. Just testing some edge cases to be sure. Looks good!

DiegoPino commented:
Hey @MorganDawe. So tested with a lot of strange characters and combinations to be sure it does not fail. It (almost) consistently escapes every character defined and facets + breadcrumbs are fine.

Some notes: 

**A) about defaults**
1. You enable escaping for advanced search
2. If you happen to remove the default regex (empty field) and you save
== Is the same as disable advanced search escaping right?

3. Since you feel it is the same, you do some searches and say: hey I will deactivate escaping.
4. Now your boss asks you to active it again: but hey: the default regex is empty. (the empty value was carried on even after deactivating/activating). 

Questions: Is that the desired behavior?


**B) about "Islandora query block"**

If I search for this with the advanced search escaping activated and default regex
RELS_EXT_hasModel_uri_s with value `info:fedora/*` i get of course 0 results
same goes for 
`"info:fedora/"*'
But Islandora Query Block displays:
```
Query

(-) RELS_EXT_hasModel_uri_s:(info:fedora/*)
```

Same as if I did that query with "escaping disabled".
being in that case previous display correct. But with "escaping enabled" should it not be really this?

```
Query 

(-) RELS_EXT_hasModel_uri_s:(info\:fedora/\*)
```

Not sure, asking.

**C) search field does not keep original value  (when the field contains `\`) and `\`is not escaped even if present in the regex.**
1. Advanced search escaping enabled
2. I have an advanced search field `dc.title`
3. I fill the field with this: `\hey` which happens to be one of my test `dc.title` entries
4. search went bad. Could not find it and also field resets itself to "hey"
Just in case tested with fgs_label if case some strange XML thing could be changing the value. Also inside the Solr index (Solr/admin..) the `\`is present as a string.
example
```XML
 <str name="fgs_label_s">\hey</str>
```

That would be my notes. Probably the first 2 are open to discussion or can be neglected, mostly about expectations and the code looks pretty good,

The third one is confusing. Since the whole idea of having escaping is to allow certain special characters that mean stuff in Solr to be handled as normal characters (present in some string). Maybe I'm testing this wrong, and maybe the backslash is really an exception, but then the regex should take care of it.

@whikloj do you have any suggestions here? pinging you also since you have been (thanks) very active on Solr.

Thanks.

MorganDawe commented:
Hi @DiegoPino ,
I have been doing some testing regarding your feedback.

A) I have tried to address this by adding the default regex as a copy past string, and added it to the reset to default functionality.
B) The query block value received the 'stripslashes()' treatment, as a quality of life improvement. Cleaning up the value for display. Like you said, its a bit of a debatable point, but reverting this would be pretty easy.

C) This last point is the most interesting one, I am finding as well that updating the regex to include backslash escaping is not function as expected, (ex: /(\+|-|&&|\|\||!|\(|\)|\{|\}|\[|\]|\^|~|\*|\?|\:|\|\\\\")/) So i am going to test that a little further as well. However, i was not getting a broken search like a 400 error.

MorganDawe committed:
Corrected variable name.

MorganDawe committed:
Merge pull request #7 from MorganDawe/MorganDawe-7.x-1889

Corrected variable name.

MorganDawe commented:
@DiegoPino My mistake. Upon closer inspection one of the variable names was incorrect, and was preventing a properly configured regex field from being escaped, such as `\`. With this correction in mind, i have tested again using the following regex (configured in UI):

`/(\+|-|&&|\|\||!|\(|\)|\{|\}|\[|\]|\^|~|\*|\?|\:|\"|\\)/`

I have also created a new object with a label of `\hey`, and added `fgs_label_s`, or `<str name="fgs_label_s">\hey</str>` as it appears in the solr doc and preformed an advanced search on `fgs_label_s` value of `\hey`. Search results now return the expected result, and is properly escaping the `\` character in the query without error. 

```
[start] => 0
[q] => fgs_label_s:(\\hey)
```
I have updated this pull request with the proper variable name, so escaping should work as intended now.

Thank you for the review, I feel silly for missing that 👍. I am eager to get your thoughts on this latest update.

DiegoPino commented:
@MorganDawe no worries, it's a long and sinuous road this escaping thing. I can retest this sadly only tomorrow. Hope that is fine for you. Cheers

MorganDawe committed:
Added \ to list of escape chars.

MorganDawe committed:
Merge pull request #8 from MorganDawe/MorganDawe-7.x-1889

Added \ to list of escape chars.

MorganDawe commented:
@DiegoPino Please let me know when you get a chance to take a look at the latest code. Thank you.

DiegoPino commented:
@MorganDawe will start again, I was testing just before your last two commits came. Give me today please. Thanks

DiegoPino commented:
@MorganDawe did the same tests and C) is solved. We did not get more interaction from other @Islandora/7-x-1-x-committers so I'm assuming the current A) and B) ways of handling escaping in UI are the expected ways of doing things, and if not, we would have to go for another pull request. Thanks a lot, will merge now.
